addons/website_sale_autocomplete/controllers/main.py

```diff
@@ -0,0 +1,151 @@
+                   'administrative_area_level_1', 'administrative_area_level_2']
+GOOGLE_PLACES_ENDPOINT = 'https://maps.googleapis.com/maps/api/place'
+TIMEOUT = 2.5
+
+
+class AutoCompleteController(http.Controller):
+
+    def _get_key(self):
+        return request.env['website'].get_current_website().google_places_api_key
+
+    def _translate_google_to_standard(self, google_fields):
+        standard_data = {}
+
+        for google_field in google_fields:
+            fields_standard = FIELDS_MAPPING[google_field['type']] if google_field['type'] in FIELDS_MAPPING else []
+
+            for field_standard in fields_standard:
+                if field_standard in standard_data:  # if a value is already assigned, do not overwrite it.
+                    continue
+                # Convert state and countries to odoo ids
+                if field_standard == 'country':
+                    standard_data[field_standard] = request.env['res.country'].search(
+                        [('code', '=', google_field['short_name'].upper())])[0].id
+                elif field_standard == 'state':
+                    state = request.env['res.country.state'].search(
+                        [('code', '=', google_field['short_name'].upper()),
+                         ('country_id.id', '=', standard_data['country'])])
+                    if len(state) == 1:
+                        standard_data[field_standard] = state.id
+                else:
+                    standard_data[field_standard] = google_field['long_name']
+        return standard_data
+
+    def _guess_number_from_input(self, source_input, standard_address):
+        """
+        Google might not send the house number in case the address
+        does not exist in their database.
+        We try to guess the number from the user's input to avoid losing the info.
+        """
+        # Remove other parts from address to make better guesses
+        guess = source_input.replace(standard_address.get('zip', ''), '') \
+            .replace(standard_address.get('street', ''), '') \
+            .replace(standard_address.get('city', ''), '')
+        guess = guess.split(',')[0].strip()
+        return guess
+
+    def _perform_place_search(self, partial_address, session_uuid=None, language_code=None, country_code=None):
+        params = {
+            'key': self._get_key(),
+            'fields': 'formatted_address,name',
+            'inputtype': 'textquery',
+            'types': 'address',
+            'input': partial_address
+        }
+        if country_code:
+            params['components'] = f'country:{country_code}'
+        if language_code:
+            params['language'] = language_code
+        if session_uuid:
+            params['sessiontoken'] = session_uuid
+
+        try:
+            results = requests.get(f'{GOOGLE_PLACES_ENDPOINT}/autocomplete/json', params=params, timeout=TIMEOUT).json()
```

Comment:
@odony Sorry about that, here it is :

Here is the first `requests` call. I think this should be safe according to what you said.
- `GOOGLE_PLACES_ENDPOINT` is a constant
- For the params : `partial_address`is the text entered by the user, `country_code`and `language_code` could also be messed with by the user.
- I wonder about `session_uuid`, it cannot be set by the user, but it contains `request.session.sid`. I don't know whether that's fine or not.
- The call has a timeout and the exception is caught (Added after reading your comment)

Reply:
Does the places API never error? Because here there's basically no error handling aside from the timeout.

If the API can send non-200 responses and the content is not JSON it'll blow up here, and if it's a JSON without a `predictions` key it'll break at line 86.

Reply:
> I wonder about session_uuid, it cannot be set by the user, but it contains `request.session.sid`. I don't know whether that's fine or not.

Why are you sending our session sid? The Places docs say:

>  The session begins when the user starts typing a query, and concludes when they select a place and a call to Place Details is made. Each session can have multiple autocomplete queries, followed by one place selection.

That's not how browser sessions work, the session gets destroyed after 90days of inactivity by default, that doesn't match Google's guidance.

Plus it's not a UUID so the name of the local variable just wrong.